agent/consul/state/state_store.go

```diff
@@ -84,10 +84,7 @@ const (
 	// higher-level watch that's less fine-grained. This isn't as bad as it
```

Comment:
This comment about it "not being as bad as it seems" turns out to be the core bad assumption here - most users who have enough scale to hit 2k nodes in a blocking query also have enough scale that they have near constant churn of instances/KV values and so the fallback to a more-frequently increasing value turns all the expensive watches into hot loops that return full responses almost constantly.

Perhaps we should update this comment?

I just tried for a bit to se if increasing `aFew` would help reduce goroutines but it's all just too complex - higher values of `aFew` end up benchmarking much slower in pure "watching" tasks because Go `select` statements are linear in number of cases. So it's hard to know what to optimize for here - more goroutines on servers increases CPU scheduler pressure and memory and GC etc. but they are still pretty cheap compared with triggering the whole watch to return too often because we fell back to a higher-level index.

Everything I could think of ended up significantly changing how WatchSets and MemDB work and I've already gone down that road and the end result is streaming which is coming 🔜 so I think this for now.

Maybe just change the comment to reflect this uncertainty/temporaryness and link to this PR or the issue above if people want to see the history?

Reply:
Aggred! I updated the comment.